File: packages/sdks/server-sdks/gatrix-dotnet-server-sdk/src/Gatrix.Server.Sdk/Evaluation/Strategies/UserWithIdStrategy.cs
namespace Gatrix.Server.Sdk.Evaluation.Strategies;

/// <summary>
/// User with ID strategy — enables for specific user IDs.
/// </summary>
public class UserWithIdStrategy : IStrategy
{
    public string Name => "userWithId";

    public bool IsEnabled(Models.StrategyParameters? parameters, Models.EvaluationContext context, string flagName)
    {
        var userIds = parameters?.UserIds;
        if (string.IsNullOrEmpty(userIds)) return false;
        if (string.IsNullOrEmpty(context.UserId)) return false;

        var ids = userIds.Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries);
        return ids.Contains(context.UserId, StringComparer.Ordinal);
    }
}

File: packages/evaluator/src/strategies/user-with-id-strategy.ts
/**
 * User With ID Strategy
 *
 * Enables feature for specific user IDs listed in parameters.userIds.
 * No stickiness - direct userId match.
 */

import {
  EvaluationContext,
  StrategyParameters,
  StrategyEvaluationResult,
} from '@gatrix/shared';
import { Strategy } from './strategy';

export class UserWithIdStrategy extends Strategy {
  constructor() {
    super('userWithId');
  }

  isEnabled(
    parameters: StrategyParameters,
    context: EvaluationContext
  ): boolean {
    const userIds = parameters.userIds;
    if (!userIds || !context.userId) return false;
    const matched = userIds.split(/\s*,\s*/).includes(context.userId);
    return matched;
  }

  isEnabledWithDetails(
    parameters: StrategyParameters,
    context: EvaluationContext
  ): StrategyEvaluationResult {
    const userIds = parameters.userIds;
    if (!userIds) {
      return { enabled: false, reason: 'No userIds defined in parameters' };
    }

    if (!context.userId) {
      return {
        enabled: false,
        reason: 'No userId in context',
        details: { userIds },
      };
    }

    const userIdList = userIds.split(/\s*,\s*/);
    const matched = userIdList.includes(context.userId);

    return {
      enabled: matched,
      reason: matched
        ? `userId "${context.userId}" found in list`
        : `userId "${context.userId}" not in list [${userIdList.join(', ')}]`,
      details: { contextUserId: context.userId, userIdList },
    };
  }
}
